pkg/agent/multicluster/stretched_networkpolicy_controller.go

```diff
@@ -0,0 +1,424 @@
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	podLabel := s.podToLabel[oldPodRef]
+	s.deleteLabelToPod(podLabel, oldPodRef)
+	delete(s.podToLabel, oldPodRef)
+}
+
+// processNamespaceUpdate handles Namespace update events. It only enqueues all
+// Pods in this Namespace if the Labels of this Namespace has been updated.
+func (s *StretchedNetworkPolicyController) processNamespaceUpdate(old, cur interface{}) {
+	oldNS, oldIsNS := old.(*v1.Namespace)
+	curNS, curIsNS := cur.(*v1.Namespace)
+	if !oldIsNS || !curIsNS {
+		klog.ErrorS(fmt.Errorf("received unexpected object"), "StretchedNetworkPolicyController Namespace UpdateFunc can't process event", "oldObj", old, "curObj", cur)
+		return
+	}
+	if reflect.DeepEqual(oldNS.Labels, curNS.Labels) {
+		klog.V(2).InfoS("StretchedNetworkPolicyController Namespace UpdateFunc received a Namespace update event, but labels are the same. Skip it", "Namespace", curNS.Name)
+		return
+	}
+	allPodsInNS, err := s.podLister.Pods(curNS.Name).List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "StretchedNetworkPolicyController Namespace UpdateFunc can't list Pods in the Namespace", "Namespace", curNS.Name)
+	}
+	for _, pod := range allPodsInNS {
+		if pod.Spec.NodeName != s.nodeName {
+			continue
+		}
+		normalizedLabel := multicluster.GetNormalizedLabel(curNS.Labels, pod.Labels, curNS.Name)
+		classifierFlowUpdate := s.getClassifierFlowUpdate(getPodReference(pod), normalizedLabel)
+		s.queue.Add(classifierFlowUpdate)
+	}
+}
+
+// processLabelIdentityAddOrUpdate handles labelIdentity add or update event. It will
+// 1. Add/Update labelIdentity to labelIdentityCache
+// 2. Enqueue all Pods with this label Identity
+func (s *StretchedNetworkPolicyController) processLabelIdentityAddOrUpdate(obj interface{}) {
+	labelIdentity, isLabelID := obj.(*v1alpha1.LabelIdentity)
+	if !isLabelID {
+		klog.ErrorS(fmt.Errorf("received unexpected object"), "StretchedNetworkPolicyController LabelIdentity AddOrUpdateFunc can't process event", "obj", obj)
+		return
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.labelIdentityCache[labelIdentity.Spec.Label] = labelIdentity.Spec.ID
+	if podSet, ok := s.labelToPods[labelIdentity.Spec.Label]; ok {
+		for podRef := range podSet {
+			classifierFlowUpdate := s.getClassifierFlowUpdateLocked(podRef, s.podToLabel[podRef])
+			s.queue.Add(classifierFlowUpdate)
+		}
+	}
+}
+
+func (s *StretchedNetworkPolicyController) processLabelIdentityDelete(obj interface{}) {
+	labelIdentity, isLabelID := obj.(*v1alpha1.LabelIdentity)
+	if !isLabelID {
+		klog.ErrorS(fmt.Errorf("received unexpected object"), "StretchedNetworkPolicyController LabelIdentity DeleteFunc can't process event", "obj", obj)
+		return
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	delete(s.labelIdentityCache, labelIdentity.Spec.Label)
+	if podSet, ok := s.labelToPods[labelIdentity.Spec.Label]; ok {
+		for podRef := range podSet {
+			classifierFlowUpdate := s.getClassifierFlowUpdateLocked(podRef, s.podToLabel[podRef])
```

Comment:
I am wondering can we guarantee the sequence?
a pod changes its label which cause a LabelIdentity delete event, and a new labelIdentity is assigned. A pod event update will be triggered and LabelIdentity delete/add event will happen, if different workers handle these events, will it happen that old id overrides new id?

Reply:
No matter what sequence, the last event is either labelIdentity add or pod update. When processing the last event, everything should have already been updated to the latest version.